From 14b8da8b355d9a55f04602ccf10763a680bf2989 Mon Sep 17 00:00:00 2001
From: Phillip Lougher <phillip@squashfs.org.uk>
Date: Mon, 14 Jul 2014 01:01:26 +0100
Subject: [PATCH 014/120] actions: use access() as an initial validity check in
 contained_followlink()

if access() returns error then the symlink cannot be dereferenced
for some reason, and it is pointless walking the symlink because it
will fail.  This also deals with cases where a symbolic link is
circular, generating -ELOOP, and so we don't have to worry about
dealing with loops ourselves.

Signed-off-by: Phillip Lougher <phillip@squashfs.org.uk>
---
 squashfs-tools/action.c     | 8 ++++++++
 squashfs-tools/mksquashfs.c | 2 +-
 2 files changed, 9 insertions(+), 1 deletion(-)

diff --git a/squashfs-tools/action.c b/squashfs-tools/action.c
index 363c73c..266bc08 100644
--- a/squashfs-tools/action.c
+++ b/squashfs-tools/action.c
@@ -2482,6 +2482,14 @@ static int contained_followlink_fn(struct atom *atom,
 	if(action_data->dir_ent->nonstandard_pathname)
 		return 0;
 
+	/* if access() returns error then the symlink cannot be dereferenced
+	 * for some reason, and it is pointless walking the symlink because it
+	 * will fail.  This also deals with cases where a symbolic link is
+	 * circular, generating -ELOOP, and so we don't have to worry about
+	 * dealing with loops ourselves */
+	if(access(action_data->pathname, F_OK) == -1)
+		return 0;
+
 	bytes = readlink(action_data->pathname, buff, 65536);
 	if(bytes < 1 || bytes == 65536)
 		/* reading symlink failed or (unlikely) the symlink was longer
diff --git a/squashfs-tools/mksquashfs.c b/squashfs-tools/mksquashfs.c
index 9a662ee..cfb4fcb 100644
--- a/squashfs-tools/mksquashfs.c
+++ b/squashfs-tools/mksquashfs.c
@@ -4985,7 +4985,7 @@ void calculate_queue_sizes(int mem, int *readq, int *fragq, int *bwriteq,
 
 
 #define VERSION() \
-	printf("mksquashfs version 4.3-git (2014/06/22)\n");\
+	printf("mksquashfs version 4.3-git (2014/07/13)\n");\
 	printf("copyright (C) 2014 Phillip Lougher "\
 		"<phillip@squashfs.org.uk>\n\n"); \
 	printf("This program is free software; you can redistribute it and/or"\
-- 
2.8.3

